types/json_path_expr.go

```diff
@@ -50,34 +50,89 @@ import (
 		select json_extract('{"a": "b", "c": [1, "2"]}', '$.*') -> ["b", [1, "2"]]
 */
 
+// if index is positive, it represents the [index]
+// if index is negative, it represents the [len() + index]
+// a normal index "5" will be parsed into "5", and the "last - 5" will be "-6"
+type jsonPathArrayIndex int
+
+func (index jsonPathArrayIndex) getIndexFromStart(obj BinaryJSON) int {
+	if index < 0 {
+		return obj.GetElemCount() + int(index)
+	}
+
+	return int(index)
+}
+
+// validateIndexRange returns whether `a` could be less or equal than `b`
+// if two indexes are all non-negative, or all negative, the comparison follows the number order
+// if the sign of them differs, this function will still return true
+func validateIndexRange(a jsonPathArrayIndex, b jsonPathArrayIndex) bool {
+	if (a >= 0 && b >= 0) || (a < 0 && b < 0) {
+		return a <= b
+	}
+
+	return true
+}
+
+func jsonPathArrayIndexFromStart(index int) jsonPathArrayIndex {
+	return jsonPathArrayIndex(index)
+}
+
+func jsonPathArrayIndexFromLast(index int) jsonPathArrayIndex {
+	return jsonPathArrayIndex(-1 - index)
+}
+
+type jsonPathArraySelection interface {
+	jsonPathArraySelection()
```

Comment:
this function is never called, and I see a lot of type assert in `extractTo`, to get the index range, why not set a function like 
```go
getIndexRange() (start, end)
```

we can make `start` = `end` to indicate the single index and uniform the all

Reply:
We at least will need to distinguish `[*]` and the others (as `-1` represents `last`, we cannot use it as asterisk now).

Another problem is to implement `String()`. If we uniform them all with `start = end`, the `[2]` after parse and `String()` will be serialized as `[2 to 2]` (or `[2 to 2]` will become `[2]`, we cannot distinguish them). But it's not correct, as it loses the information carried by flags (`containsRange`), `json_extract('[1,2,3]', '$[1]')` will return `2`, while `json_extract('[1,2,3]', '$[1 to 1]')` will give `[2]`.